### ISSUE TITLE
Running decoders in parallel

### ISSUE THREAD

**Issue by DeepHorizons** [bambocher/pocketsphinx-python#21]
I'm trying to run decoders in parallel for an application I'm writing. It seems that no matter what I do, it takes the same amount of time whether I'm using multiple decoders or just one. Doing a profile shows that almost all the time is spent acquiring a lock.

Are there are suggestions for running decoders in parallel?

**Comment by nshmyrev** [comment]
What kind of lock? There are no locks in current code. And you should not use them either.

**Comment by DeepHorizons** [comment]
I'm not using locks, it appears that it is in one of the compiled files, maybe in sphinxbase?

Here is a sample script that I am running
```python
import cProfile
import threading
import pocketsphinx

model_path = pocketsphinx.get_model_path()
data_path = pocketsphinx.get_data_path()

p = cProfile.Profile()

# Creating the decoders first only saves a little time, but still does not run in parallel
#decoders = [pocketsphinx.Pocketsphinx(verbose=False) for _ in range(5)]

def test(n):
    print('starting thread {}'.format(n))
    decoder = pocketsphinx.Pocketsphinx(verbose=False)
    #decoder = decoders[n]
    print('{} created decoder {}'.format(n, repr(decoder)))
    result = str(decoder.decode())
    print('text: {}'.format(result))

def run_threads(n, func=test):
    threads = [threading.Thread(target=func, args=(n,)) for n in range(n)]
    print('created all threads')
    p.enable()
    [t.start() for t in threads]
    print('started all threads')
    [t.join() for t in threads]
    p.disable()
    p.print_stats(sort=1)
    p.clear()
    print('joined all threads')
```

And here is the output of the profile when running `run_threads(5)`
```
         148 function calls in 20.164 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       25   18.593    0.744   18.593    0.744 {method 'acquire' of '_thread.lock' objects}
        1    1.569    1.569    1.569    1.569 {built-in method builtins.print}
        5    0.001    0.000    0.001    0.000 {built-in method _thread.start_new_thread}
        1    0.000    0.000   14.826   14.826 pocketsphinx_time_test.py:27(<listcomp>)
        5    0.000    0.000   14.826    2.965 threading.py:1022(join)
        5    0.000    0.000   14.825    2.965 threading.py:1060(_wait_for_tstate_lock)
        5    0.000    0.000    3.769    0.754 threading.py:531(wait)
        5    0.000    0.000    3.769    0.754 threading.py:261(wait)
        5    0.000    0.000    3.770    0.754 threading.py:826(start)
        5    0.000    0.000    0.000    0.000 threading.py:964(_stop)
        5    0.000    0.000    0.000    0.000 {built-in method _thread.allocate_lock}
        5    0.000    0.000    0.000    0.000 threading.py:1224(current_thread)
        1    0.000    0.000    3.770    3.770 pocketsphinx_time_test.py:25(<listcomp>)
        5    0.000    0.000    0.000    0.000 threading.py:249(_acquire_restore)
        5    0.000    0.000    0.000    0.000 threading.py:240(__exit__)
        5    0.000    0.000    0.000    0.000 threading.py:237(__enter__)
        2    0.000    0.000    0.000    0.000 cp437.py:18(encode)
        5    0.000    0.000    0.000    0.000 threading.py:252(_is_owned)
        2    0.000    0.000    0.000    0.000 {built-in method _codecs.charmap_encode}
        5    0.000    0.000    0.000    0.000 threading.py:246(_release_save)
       10    0.000    0.000    0.000    0.000 {method 'release' of '_thread.lock' objects}
       10    0.000    0.000    0.000    0.000 threading.py:504(is_set)
        5    0.000    0.000    0.000    0.000 {method '__enter__' of '_thread.lock' objects}
        5    0.000    0.000    0.000    0.000 {built-in method _thread.get_ident}
        5    0.000    0.000    0.000    0.000 {method 'locked' of '_thread.lock' objects}
        5    0.000    0.000    0.000    0.000 {method 'append' of 'collections.deque' objects}
        5    0.000    0.000    0.000    0.000 {method '__exit__' of '_thread.lock' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

Running it with one thread takes about 4 seconds, 5 threads takes about 20 seconds.

**Comment by nshmyrev** [comment]
Well, explained here:

http://stackoverflow.com/questions/18114285/python-what-are-the-differences-between-the-threading-and-multiprocessing-modul

You'd better use mulitprocessing. Though some code to release GIL inside pocketsphinx is possible too.

**Comment by bambocher** [comment]
Fixes 6f94efe
